File: Drivers/FSCC Drivers/linux/lib/c++/utils/loop/loop.cpp
#include <conio.h>
#include <iostream>

#include <Windows.h>

#include <fscc.hpp>

#define DATA_LENGTH 20

int init(Fscc::Port &p);
int loop(Fscc::Port &p1, Fscc::Port &p2);

int main(int argc, char *argv[])
{
    int e = 0;
    unsigned port_num_1, port_num_2;
    BOOL reset;
    unsigned iterations = 0;
    unsigned mismatched = 0;

    if (argc < 3 || argc > 4) {
        fprintf(stdout, "%s PORT_NUM PORT_NUM [RESET_REGISTER=1]", argv[0]);
        return EXIT_FAILURE;
    }

    port_num_1 = atoi(argv[1]);
    port_num_2 = atoi(argv[2]);
    reset = (argc == 4) ? atoi(argv[3]) : 1;

    Fscc::Port p1((unsigned)port_num_1);
    Fscc::Port p2((unsigned)port_num_2);

    if (reset) {
        e = init(p1);
        if (e != 0)
            return EXIT_FAILURE;
        
        e = init(p2);
        if (e != 0)
            return EXIT_FAILURE;
    }

    std::cout << "Data looping, press any key to stop..." << std::endl;

    while (_kbhit() == 0) {
        e = loop(p1, p2);
        if (e != 0) {
            if (e == ERROR_INVALID_DATA)
                mismatched++;
            else
                return EXIT_FAILURE;
        }

        iterations++;
    }

    if (mismatched == 0)
        fprintf(stdout, "Passed (%d iterations).", iterations);
    else
        fprintf(stderr, "Failed (%d out of %d iterations).", 
                mismatched, iterations);

    return EXIT_SUCCESS;
}

int init(Fscc::Port &p)
{
    Fscc::Registers r;
    Fscc::MemoryCap m;

    m.input = 1000000;
    m.output = 1000000;

    p.SetMemoryCap(m);

    p.DisableRxMultiple();
    p.DisableAppendStatus();
    p.DisableAppendTimestamp();
    p.SetTxModifiers(Fscc::TxModifiers::XF);
    p.DisableIgnoreTimeout();

    r.FIFOT = 0x08001000;
    r.CCR0 = 0x0011201c;
    r.CCR1 = 0x00000018;
    r.CCR2 = 0x00000000;
    r.BGR = 0x00000000;
    r.SSR = 0x0000007e;
    r.SMR = 0x00000000;
    r.TSR = 0x0000007e;
    r.TMR = 0x00000000;
    r.RAR = 0x00000000;
    r.RAMR = 0x00000000;
    r.PPR = 0x00000000;
    r.TCR = 0x00000000;
    r.IMR = 0x0f000000;
    r.DPLLR = 0x00000004;
    r.FCR = 0x00000000;

    p.SetRegisters(r);

    p.SetClockFrequency(18432000);
    
    p.Purge(true, true);

    return ERROR_SUCCESS;
}

int loop(Fscc::Port &p1, Fscc::Port &p2)
{
    unsigned bytes_written = 0, bytes_read = 0;
    char odata[DATA_LENGTH];
    char idata[100];

    memset(odata, 0x01, sizeof(odata));
    memset(&idata, 0, sizeof(idata));

    bytes_written = p1.Write(odata, sizeof(odata));
    bytes_read = p2.Read(idata, sizeof(idata), 1000);
    
    if (bytes_read == 0 || memcmp(odata, idata, sizeof(odata)) != 0)
        return ERROR_INVALID_DATA;

    return ERROR_SUCCESS;
}

File: Drivers/FSCC Drivers/linux/lib/python/utils/loop/loop.py
import sys
import fscc
import msvcrt


def init(p):
    print('Restoring to default settings.')

    p.memory_cap.input = 1000000;
    p.memory_cap.output = 1000000;

    p.rx_multiple = False;
    p.append_status = False;
    p.append_timestamp = False;
    p.tx_modifiers = fscc.XF;
    p.ignore_timeout = False;

    p.registers.FIFOT = 0x08001000
    p.registers.CCR0 = 0x0011201c
    p.registers.CCR1 = 0x00000018
    p.registers.CCR2 = 0x00000000
    p.registers.BGR = 0x00000000
    p.registers.SSR = 0x0000007e
    p.registers.SMR = 0x00000000
    p.registers.TSR = 0x0000007e
    p.registers.TMR = 0x00000000
    p.registers.RAR = 0x00000000
    p.registers.RAMR = 0x00000000
    p.registers.PPR = 0x00000000
    p.registers.TCR = 0x00000000
    p.registers.IMR = 0x0f000000
    p.registers.DPLLR = 0x00000004
    p.registers.FCR = 0x00000000

    p.clock_frequency = 18432000

    p.purge(True, True)

def loop(p1, p2):
    odata = b'Hello world!';

    p1.write(odata);
    idata = p2.read()[0];
    
    if (len(idata) == 0 or odata != idata):
        return -1;

    return 0;

if __name__ == '__main__':
    port_num_1, port_num_2 = 0, 0
    iterations = 0
    mismatched = 0
    
    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print('loop.exe PORT_NUM PORT_NUM [RESET_REGISTER=1]')
        exit(1)

    port_num_1, port_num_2 = sys.argv[1], sys.argv[2]
    
    try:
        if len(sys.argv) == 3:
            reset = True if sys.argv[3] > 0 else False
        else:
            reset = True
    except:
        reset = True

    p1 = fscc.Port(port_num_1)
    p2 = fscc.Port(port_num_2)
    
    if reset:
        init(p1)
        init(p2)

    print('Data looping, press any key to stop...')

    while msvcrt.kbhit() is 0:
        e = loop(p1, p2)
        if e != 0:
            if e == -1:
                mismatched += 1
            else:
                exit(1)
                
        iterations += 1

    if mismatched == 0:
        print('Passed ({} iterations).'.format(iterations))
    else:
        print('Failed ({} out of {} iterations).'.format(mismatched, iterations))

    exit(0)